Sources/Plasma/NucleusLib/pnAsyncCoreExe/pnAceSocket.cpp

```diff
@@ -0,0 +1,545 @@
+
+        asio::error_code err;
+        op.fSock.cancel(err);  // This will wake up async_connect()
+        if (err)
+            LogMsg(kLogError, "Failed to cancel socket connect operation: {}", err.message());
+    }
+}
+
+void AsyncSocketDelete(AsyncSocket conn)
+{
+    delete conn;
+}
+
+static void HardCloseSocket(AsyncSocket conn)
+{
+    asio::error_code err;
+    conn->fSock.set_option(asio::socket_base::linger(true, 0), err);
+    if (err)
+        LogMsg(kLogError, "Failed to set socket linger option: {}", err.message());
+
+    conn->fSock.close(err);
+    if (err)
+        LogMsg(kLogError, "Failed to close socket: {}", err.message());
+}
+
+void AsyncSocketDisconnect(AsyncSocket conn, bool hardClose)
+{
+    if (!conn->fSock.is_open())
+        return;
+
+    asio::error_code err;
+    conn->fSock.cancel(err);
+    if (err)
+        LogMsg(kLogError, "Failed to cancel pending operations: {}", err.message());
+
+    hsAssert(hardClose, "Only hardClose supported on sockets");
+    HardCloseSocket(conn);
+    conn->closeTimeMs |= 1;
+}
+
+static bool SocketQueueAsyncWrite(AsyncSocket conn, const void* data, size_t bytes)
+{
+    hsLockGuard(s_connectCrit);
+
+    // check for data backlog
+    WriteOperation* firstQueuedWrite = conn->fWriteOps.front();
+    if (firstQueuedWrite) {
+        unsigned currTimeMs = TimeGetMs();
+        if (((long)(currTimeMs - firstQueuedWrite->queueTimeMs) >= (long)kBacklogFailMs) && ((long)(currTimeMs - conn->initTimeMs) >= (long)kBacklogInitMs))
+        {
+            PerfAddCounter(kAsyncPerfSocketDisconnectBacklog, 1);
+
+            if (conn->fConnectionType) {
+                LogMsg(
+                    kLogPerf,
+                    "Backlog, c:{} q:{}, i:{}",
+                    conn->fConnectionType,
+                    currTimeMs - firstQueuedWrite->queueTimeMs,
+                    currTimeMs - conn->initTimeMs);
+            }
+            AsyncSocketDisconnect((AsyncSocket)conn, true);
+            return false;
+        }
+    }
+
+    // If the last buffer still has space available then add data to it
+    if (!conn->fWriteOps.empty()) {
+        WriteOperation* op = conn->fWriteOps.back();
+        size_t          bytesLeft = std::min(op->fAllocSize - op->fNotify.bytes, bytes);
+        if (bytesLeft) {
+            PerfAddCounter(kAsyncPerfSocketBytesWaitQueued, bytesLeft);
+            memcpy(op->fNotify.buffer + op->fNotify.bytes, data, bytesLeft);
+            op->fNotify.bytes += bytesLeft;
+            data = (const uint8_t*)data + bytesLeft;
+            bytes -= bytesLeft;
+        }
+    }
+
+    if (bytes) {
+        // Allocate storage alongside the operation structure itself
+        size_t          bytesAlloc = std::max(bytes, kMinBacklogBytes);
+        auto            membuf = new uint8_t[sizeof(WriteOperation) + bytesAlloc];
+        WriteOperation* op = new (membuf) WriteOperation;
+        conn->fWriteOps.emplace_back(op);
+
+        // TODO: asyncId

```

Comment:
Is this still `TODO`?

Reply:
Let me look. Async ID is not actually used anywhere. Let me see what it was for originally in context though.

Reply:
So... here's my take on Async ID. And someone might know more than me.

It looks like all Async ID was used for was tracking write operations sequentially. And they went in a linked list, so it may have just been ended up using to identify what was a write operation.

The new code has a dedicated write operation vector, so we don't need the Async ID to determine what was a write operation. Beyond that - I can't find any use for Async ID - and it probably should be removed.

I'm not sure if maybe back in the day people thought this should be connected to logging? Maybe there was some diagnostic tool that used it?